rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,374 @@
+   using `RefCell` moves the borrow checking to the runtime rather than compile
+   time. Affected areas of code will need to be very vigorously tested to avoid
+   runtime panics. Fortunately a lot of Vrl relies on creating new values
+   rather than mutating existing, so there isn't too much code that will be
+   affected.
+6. `Value` is no longer `Send + Sync`. There are some nodes in the AST that
+   need to store a `Value`, for example `Variable` stores a `Value`.  We need
+   to identify if this `Value` is actually necessary (it's not clear to me).
+   If it isn't we can remove it.
+   If it is we need to create a thread safe variable of `Value` that can be
+   stored here and then converted into a `Value` at runtime.
+
+Initial experiments roughly showed a reduction from 1m20s to 1m02s to push
+100,000 records through Vector using reference counting.
+
+### Bytecode VM
+
+Vrl compiles to an AST that is then walked during resolution. Each node in that
+tree is boxed and stored in disparate regions of memory. As a result walking
+the tree means that the CPU caches must be constantly swapped.
+
+Instead we can create a bytecode VM to store the execution of the Vrl program.
+
+Bytecode is essentially a big enum of instructions:
+
+```rust
+#[derive(FromPrimitive, ToPrimitive, Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OpCode {
+    Return = 255,
+    Constant,
+    Negate,
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Print,
+    Not,
+    Greater,
+    GreaterEqual,
+    Less,
+    LessEqual,
+    NotEqual,
+    Equal,
+    Pop,
+    JumpIfFalse,
+    Jump,
+    SetPath,
+    GetPath,
+    Call,
+    ...
+}
+```
+
+The Vm is a struct comprising of the following fields:
+
+```rust
+#[derive(Clone, Debug, Default)]
+pub struct Vm {
+    instructions: Vec<usize>,
+    values: Vec<Literal>,
+    targets: Vec<Variable>,
+    stack: Vec<Value>,
+    ip: usize,
+}
+```
+
+- instructions
+
+The instructions field is a `Vec` of `OpCode` cast to a usize. The reason for
+the cast is because not all instructions are `OpCode`. For example the
+instructions `[.., Constant, 12, ..]` when evaluated will load the constant
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- values
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- ip
+
+The instruction pointer points to the next instruction to evaluate.
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
```

Comment:
Not sure I fully understand. How so?

Reply:
When traversing the current tree representation, the jump instructions will be in a number of different places in code. Each jump point will have a separate branch prediction buffer in the CPU. This separation allows the CPU to predict the branches more often, since each jump will tend to go to only a predictable number of other locations. With byte code, there will be an opcode dispatch that becomes effectively unpredictable due to data dependencies.

I may, of course, be missing how many jumps are involved in traversing the tree, and this may well come out to be a win. I just wanted to point out a potential cost that went along with the cache efficiency win.